### ISSUE TITLE
Tuya TS0601 _TZE200_pw7mji0l Cover Motor: Open and Close state not working properly in Home Assistant

### ISSUE THREAD

**Issue by didi3r** [Koenkk/zigbee2mqtt#19665]
### What happened?

The "open" and "closed" states reported to Home Assistant are not working as expected. They seem to be based on whatever you set on the "state". 

1. If you set it to "Open" in zigbee2mqtt, and then change the position through zigbee2mqtt or the remote. The state would be "open" regarding the position.

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/2e9261ea-19a8-4694-923a-fe9a9d003833)

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/9c295a8e-4109-4499-9384-db63bd03111d)

1. If you set it to "Close" in zigbee2mqtt, and then change the position through zigbee2mqtt or the remote. The state would be "closed" regarding the position.

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/db4e4a69-ef52-458f-894f-6ec18a5e1eb2)

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/f4e4029f-8ea6-4b1c-9b17-e3740c1a9cab)

3. If you set it to "Stop" in zigbee2mqtt, and then change the position through zigbee2mqtt or the remote. The state would be "open" or "closed" **based on the position**.

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/cde98def-72fa-40ec-a0c6-82e7b76b28ee)

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/8c99b138-0a12-4c3f-bc35-12841cb756a5)

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/9f79d9e2-4b85-40a2-891a-a585fc8fd0ce)

![image](https://github.com/Koenkk/zigbee2mqtt/assets/2186020/193052ff-4946-4445-ba9e-35f04c4c2df4)

### What did you expect to happen?

I would like the reported state of the covers to be always based on the position, not on what the user can randomly change.

### Zigbee2MQTT version

1.33.2-dev commit: c8dcd71

### Adapter firmware version

20220219

### Adapter

Sonoff ZBDongleP (CC2652P)

**Comment by GvdbW** [comment]
I have the same issue with the roller TZE200_bv1jcqqu

it all worked normally until the last update. Now all my automation are broken and work randomly. 

It seems related to this issue 
https://github.com/Koenkk/zigbee2mqtt/issues/19539

**Comment by elad-bar** [comment]
i have the same issue for all tuya homeetec covers, checked the converters (i added them almost a year ago), nothing change in them but it now throws errors as described above

**Comment by elad-bar** [comment]
adding more details, devices:
- _TZE200_p6vz3wzt
- _TZE200_jhkttplm
- _TZE200_5nldle7w

**Comment by alexeiw123** [comment]
I am seeing this behaviour as well, has anyone identified the source of the problem?

**Comment by alexeiw123** [comment]
@Koenkk how can we take an integrated device from .ts and use that to in a local .js external converter? I'd like to see if I can find a way around this but the only way I can see is starting from scratch.

The devices position updates OK so somehow just need the state to reflect the position regardless of if it is operated by ZigBee or RF remote.

**Comment by Koenkk** [comment]
Could you check if the issue is fixed with the following [external converter](https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4): 

- save this as file next to `configuration.yaml` as `ext_converter.js`
- add it to `configuration.yaml`:
```yaml
external_converters:
  - ext_converter.js
```
- start z2m, 
- in the frontend, check if the description of this device now ends with `CUSTOM` (this indicates the external converter has been loaded correctly)
- check if issue is fixed

**Comment by alexeiw123** [comment]
> Could you check if the issue is fixed with the following [external converter](https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4):
> 
> * save this as file next to `configuration.yaml` as `ext_converter.js`
> * add it to `configuration.yaml`:
> 
> ```yaml
> external_converters:
>   - ext_converter.js
> ```
> 
> * start z2m,
> * in the frontend, check if the description of this device now ends with `CUSTOM` (this indicates the external converter has been loaded correctly)
> * check if issue is fixed

Thanks - I have added this external converter and I get no errors in my logs, however the blind description is still `cover motor` not `cover motor CUSTOM` as I'd expect.

external converter is showing in the GUI settings, so config seems to be working OK.

**Comment by alexeiw123** [comment]
> * check if issue is fixed

OK I updated this line:
`tuya.whitelabel('Zemismart', 'ZM25EL', 'Cover motor', ['_TZE200_pw7mji0l']),`
To this:
`tuya.whitelabel('Zemismart', 'ZM25EL', 'Cover motor CUSTOM', ['_TZE200_pw7mji0l']),`

The external converter was loading but the white label line was preventing the description form updating

**Comment by alexeiw123** [comment]
> * check if issue is fixed

Unfortunately no - using the external converter, the position expose doesn't update at all anymore.

**Comment by alexeiw123** [comment]
temporary workaround:

This will create a new cover entity in HA that interacts with the Z2MQTT created cover entity, but updates state based on the reported position in Z2M. in my very short testing this seems to be working for me until the converter issue is figure out.

```
cover:
  - platform: template
    covers:
      your_blind_temp:
        device_class: shade
        friendly_name: "Your blind temp"
        unique_id: your_blind_temp
        position_template: >-
          {{ state_attr('cover.your_blind', 'current_position')}}
        value_template: >-
          {% if state_attr('cover.your_blind', 'current_position') | int(default=0) > 0 %}
            open
          {% else %}
            closed
          {% endif %}
        open_cover:
          action: cover.open_cover
          target:
            entity_id: cover.your_blind
        close_cover:
          action: cover.close_cover
          target:
            entity_id: cover.your_blind
        stop_cover:
          action: cover.stop_cover
          target:
            entity_id: cover.your_blind
        set_cover_position:
          action: cover.set_cover_position
          data:
            position: "{{position}}"
          target:
            entity_id: cover.your_blind
```

**Comment by Koenkk** [comment]
Could you provide the debug log when changing the position of the cover and it not being updated?

See [this](https://www.zigbee2mqtt.io/guide/configuration/logging.html) on how to enable debug logging.

**Comment by alexeiw123** [comment]
> Could you provide the debug log when changing the position of the cover and it not being updated?
> 
> See [this](https://www.zigbee2mqtt.io/guide/configuration/logging.html) on how to enable debug logging.

[log.log](https://github.com/user-attachments/files/16631689/log.log)

Here you go. FYI I opened the blind at 10:58 using the RF remote - I have trimmed the log from before that to reduce irrelevant content.

`[2024-08-16 10:58:12] debug: 	z2m: Received Zigbee message from 'Ava blind'` <-- this is the device I tested on

**Comment by Koenkk** [comment]
I updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, can you try again?

**Comment by alexeiw123** [comment]
> I updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, can you try again?

Unfortunately with this converter the position value doesn't get populated. The state doesn't change on cover movement either.

**Comment by Koenkk** [comment]
I updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, can you provide the logging when changing the position again?

**Comment by alexeiw123** [comment]
Have tested again, here are the logs:
[log.log](https://github.com/user-attachments/files/16652810/log.log)
closed from Z2M at 08:28 and then opened from RF remote at 8:29.

Z2M was throwing this: `z2m: Exception while calling fromZigbee converter: logger.error is not a function}`

**Comment by Koenkk** [comment]
Fixed https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4

**Comment by alexeiw123** [comment]
> Fixed https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4

[log.log](https://github.com/user-attachments/files/16688402/log.log)


I believe that these lines (2640 on) capture the blind reaching the open position from an external input, returning state open but the state in Z2M is shown as `CLOSE`.

```
[2024-08-21 18:07:22] debug: 	z2m: Received Zigbee message from 'Ava blind', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,100],"type":"Buffer"},"datatype":2,"dp":2}],"seq":256}' from endpoint 1 with groupID 0
[2024-08-21 18:07:22] error: 	z2m: RECIEVED 100
[2024-08-21 18:07:22] error: 	z2m: RETURN POSITON 100 and STATE OPEN
[2024-08-21 18:07:22] info: 	z2m:mqtt: MQTT publish: topic 'zigbee2mqtt/Ava blind', payload '{"battery":90,"border":"up","click_control":"down","linkquality":144,"motor_fault":false,"position":{"position":100,"state":"OPEN"},"reverse_direction":"back","state":"CLOSE"}'
```

Seems to have logged those values this time. Same behaviour and noticed that

**Comment by Koenkk** [comment]
Updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, can you try again?

**Comment by alexeiw123** [comment]
> Updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, can you try again?

🎉 this does appear to have fixed it for me. Do you need logs or anything from me?

**Comment by Koenkk** [comment]
I updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, could you do one more check?

**Comment by alexeiw123** [comment]
> I updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, could you do one more check?

Quick test - all seems to be working OK both with commands from Z2M, HA and RF remote.

I see the changes are in the 'invert cover' which I didn't see any issue with either in this version or previous.

I'll update here if I come across any issues but I think you've fixed this issue.

**Comment by alexeiw123** [comment]
> I updated https://gist.github.com/Koenkk/ea7f0f705ad78880088b84ac380415f4, could you do one more check?

Ran in to a weird issue on this one - HA state wasn't updating from closed to open, regardless of where I was using it from. It certainly was earlier, but all my covers and groups started showing the same after I set up a few of these devices that I hadn't unboxed yet. Restart of HA and Z2M didn't resolve. Going back to the previous example from your gist has them working fine again. (code below)

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const e = exposes.presets;
const ea = exposes.access;
const legacy = require('zigbee-herdsman-converters/lib/legacy');
const lumi = require('zigbee-herdsman-converters/lib/lumi');
const utils = require('zigbee-herdsman-converters/lib/utils');
const {logger} = require('zigbee-herdsman-converters/lib/logger');

const valueConverter = {
    coverPosition: {
        to: async (v, meta) => {
            return meta.options.invert_cover ? 100 - v : v;
        },
        from: (v, meta, options, publish) => {
            logger.error(`RECIEVED ${v}`);
            // TODO: add invert cover
            const position = v;
            const closed = position === 0;
            logger.error(`RETURN POSITON ${v} and STATE ${closed ? 'CLOSE' : 'OPEN'}`);
            publish({position, state: closed ? 'CLOSE' : 'OPEN'});
            return position;
        },
    },
};

const definition = {
    fingerprint: tuya.fingerprint('TS0601', [
        '_TZE200_68nvbio9',
        '_TZE200_pw7mji0l',
        '_TZE200_cf1sl3tj',
        '_TZE200_nw1r9hp6',
        '_TZE200_9p5xmj5r',
        '_TZE200_eevqq1uv',
    ]),
    model: 'TS0601_cover_3',
    vendor: 'Tuya',
    description: 'Cover motor CUSTOM',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    onEvent: tuya.onEventSetTime,
    options: [exposes.options.invert_cover()],
    configure: tuya.configureMagicPacket,
    exposes: [
        e.battery(),
        e.cover_position(),
        e.enum('reverse_direction', ea.STATE_SET, ['forward', 'back']).withDescription('Reverse the motor direction'),
        e.enum('border', ea.STATE_SET, ['up', 'down', 'up_delete', 'down_delete', 'remove_top_bottom']),
        e.enum('click_control', ea.STATE_SET, ['up', 'down']).withDescription('Single motor steps'),
        e.binary('motor_fault', ea.STATE, true, false),
    ],
    whiteLabel: [
        tuya.whitelabel('Zemismart', 'ZM16EL-03/33', 'Cover motor', ['_TZE200_68nvbio9']),
        tuya.whitelabel('Zemismart', 'ZM25EL', 'Cover motor local test', ['_TZE200_pw7mji0l']),
        tuya.whitelabel('Zemismart', 'ZM85EL-2Z', 'Roman Rod I type U curtains track', ['_TZE200_cf1sl3tj', '_TZE200_nw1r9hp6']),
        tuya.whitelabel('Hiladuo', 'B09M3R35GC', 'Motorized roller shade', ['_TZE200_9p5xmj5r']),
    ],
    meta: {
        // All datapoints go in here
        tuyaDatapoints: [
            [1, 'state', tuya.valueConverterBasic.lookup({OPEN: tuya.enum(0), STOP: tuya.enum(1), CLOSE: tuya.enum(2)})],
            [2, 'position', valueConverter.coverPosition],
            [3, 'position', valueConverter.coverPosition],
            [2, 'position', valueConverter.coverPosition],
            [3, 'position', valueConverter.coverPosition],
            [5, 'reverse_direction', tuya.valueConverterBasic.lookup({forward: tuya.enum(0), back: tuya.enum(1)})],
            [12, 'motor_fault', tuya.valueConverter.trueFalse1],
            [13, 'battery', tuya.valueConverter.raw],
            [
                16,
                'border',
                tuya.valueConverterBasic.lookup({
                    up: tuya.enum(0),
                    down: tuya.enum(1),
                    up_delete: tuya.enum(2),
                    down_delete: tuya.enum(3),
                    remove_top_bottom: tuya.enum(4),
                }),
            ],
            [20, 'click_control', tuya.valueConverterBasic.lookup({up: tuya.enum(0), down: tuya.enum(1)})],
        ],
    },
};

module.exports = definition;
```

**Comment by alexeiw123** [comment]
Probably not new, but saw that zemismart supply the zigbee driver for these:
https://www.zemismart.com/pages/zigbee-handler-download-38
https://drive.google.com/file/d/1yBfsBpusYI1SKvMm4o0dnYRYMmD3LNiP/view

**Comment by Koenkk** [comment]
Integrated the fix, thanks for testing!

Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now.

**Comment by alexeiw123** [comment]
> Integrated the fix, thanks for testing!
> 
> Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now.

Hi @Koenkk These devices are all working as expected but noticed that I am getting errors pop up in the Z2M UI whenever the exposes are interacted with, (for example, pressing stop, or refreshing position. I don't think this was happening when using the external converter. Any clues to what has introduced this in the logs?


```
2024-09-06 09:13:16z2m: No converter available for 'get' 'position' ()
2024-09-06 09:13:20z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
2024-09-06 09:13:58z2m: EventBus error 'Publish/mqttMessage': Cannot read properties of undefined (reading 'some')
2024-09-06 09:14:08z2m: No converter available for 'get' 'position' ()
```

**Comment by Koenkk** [comment]
Tuya devices doesn't support reading, so that's why the refresh fails. Disabled that now. `Publish/mqttMessage': Cannot read properties of undefined (reading 'some')` has already been fixed.

Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now.

**Comment by kabturek** [comment]
The fix works great! thanks for the work!

**Comment by SikaiGGMU** [comment]
Just find     _TZE204_r0jdjrvi  is experiencing the same issue. If press open, the motor will move to open,but the position does get updated (still stays at 0), and the other way around.

**Comment by juslex** [comment]
I have the same problem involving the engines
TS0601 _TZE200_fzo2pocs

**Comment by alexeiw123** [comment]
> Integrated the fix, thanks for testing!
> 
> Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now.

@Koenkk this problem that you fixed has returned since (I think) 2.0.1 release. Do you know of any commits that could have caused this?

**Comment by juslex** [comment]
Yes, here the problem came back with 2.1.

**Comment by Koenkk** [comment]
Could be fixed  by https://github.com/Koenkk/zigbee-herdsman-converters/pull/8723, can you check?

Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now and in the next release which is every 1st of the month.

**Comment by alexeiw123** [comment]
> Could be fixed by [Koenkk/zigbee-herdsman-converters#8723](https://github.com/Koenkk/zigbee-herdsman-converters/pull/8723), can you check?
> 
> Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now and in the next release which is every 1st of the month.

Yes - all working as expected on latest dev build 👏👏
